### ISSUE TITLE
tasksOrder arrays bugs: hide tasks (Dailies can't be completed!), prevent sorting tasks

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#7790]
Sometimes, the user's `tasksOrder` arrays don't contain all of the user's task IDs, or they contain IDs that shouldn't be there, causing bugs. For example:
- [ToDo list sort order is not preserved · Issue #7645](https://github.com/HabitRPG/habitrpg/issues/7645) - when IDs for completed To-Dos aren't removed from `tasksOrder.todos` - I've seen several reports of this
- Neilinha (05ccf744-331a-4ae4-8bd9-731ea3413f2e) reported damage at cron for unknown reasons. I  found that 8 of their 15 Dailies didn't have their IDs in `tasksOrder.dailys` and so they weren't showing up in the Dailies list on the website. I replicated this by deleting the IDs of some of my Dailies from my tasksOrder array. This in particular is very damaging since the user can't even tell that the Dailies exist and can't mark them complete or delete them.

Suggested actions:
- [x] make sure that the website displays all tasks even if they don't appear in the tasksOrder arrays
- [ ] make sure the mobile apps display all tasks (FYI @viirus) 
- [ ] make the Data Display Tool display all tasks (currently it doesn't; Alys will fix)
- AND/OR:
- [x] **cron does not do damage for any Dailies that are not listed in `tasksOrder.dailys`** - this is an easy partial workaround that could be done quickly to prevent the most damaging effects of this bug
- [ ] cron identifies when tasks aren't in `tasksOrder` and adds them
- [ ] cron identifies when `tasksOrder` contains IDs that should no longer be there and it deletes them
- anything else / instead of these?

**Comment by paglias** [comment]
When requesting all the tasks the taskOrder array isn't used and when requesting a single type (not used by the website) there's code to make sure all the tasks are shown

```
  // Order tasks based on tasksOrder
  if (type && type !== 'completedTodos' && type !== '_allCompletedTodos') {
    let order = (challenge || user).tasksOrder[type];
    let orderedTasks = new Array(tasks.length);
    let unorderedTasks = []; // what we want to add later

    tasks.forEach((task, index) => {
      let taskId = task._id;
      let i = order[index] === taskId ? index : order.indexOf(taskId);
      if (i === -1) {
        unorderedTasks.push(task);
      } else {
        orderedTasks[i] = task;
      }
    });

    // Remove empty values from the array and add any unordered task
    orderedTasks = _.compact(orderedTasks).concat(unorderedTasks);
    res.respond(200, orderedTasks);
  } else {
    res.respond(200, tasks);
  }
```

Probably the problem lies in the client side code then. I'd also like to make sure that at cron these inconsistencies get fixed automatically by adding missing ids and removing the non existing ones

**Comment by paglias** [comment]
Hmm in theory the client side has code to show unordered tasks too

```
      function syncUserTasks (tasks) {
        user.habits = [];
        user.todos = [];
        user.dailys = [];
        user.rewards = [];

        // Order tasks based on tasksOrder
        var groupedTasks = _(tasks)
          .groupBy('type')
          .forEach(function (tasksOfType, type) {
            var order = user.tasksOrder[type + 's'];
            var orderedTasks = new Array(tasksOfType.length);
            var unorderedTasks = []; // what we want to add later

            tasksOfType.forEach(function (task, index) {
              var taskId = task._id;
              var i = order[index] === taskId ? index : order.indexOf(taskId);
              if (i === -1) {
                unorderedTasks.push(task);
              } else {
                orderedTasks[i] = task;
              }
            });

            // Remove empty values from the array and add any unordered task
            user[type + 's'] = _.compact(orderedTasks).concat(unorderedTasks);
          }).value();
      }
```

**Comment by paglias** [comment]
From my own tests a task missing fro user.taskOrder still shows up in the list (at least on the website) and can be re-ordered correctly

**Comment by paglias** [comment]
I've deleted several ids from my own tasksOrder lists of different types and they still showed up on the website and still cannot replicate the behavior

**Comment by Alys** [comment]
Yes, sorry, my Dailies list is long, and I didn't think to look at the bottom.

**Comment by Alys** [comment]
closed in favour of #7645
